File: roller-coaster/rollercoaster.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <sys/time.h>
#include <semaphore.h>
#include <unistd.h>

#define MAX_PASSENGERS 4
#define NUM_PASSENGERS 24

//counters for how many have boarded/unboarded
int boarders;
int unboarders;

//locks to guard the counters
pthread_mutex_t mx_board;
pthread_mutex_t mx_unboard;

//semaphores to queue up and signal fully loaded/unloaded
sem_t board_queue;
sem_t unboard_queue;
sem_t load;
sem_t unload;

//Car thread
void* car_sim(void* arg)
{
	int id = *(int*)arg;
	int i;
	int served = 0;
	printf("Car%d-> start\n", id);
	
	while (served < NUM_PASSENGERS) {
		//Load the car
		printf("Car%d-> load\n", id);
		i = 0;
		while(i < MAX_PASSENGERS) {
			sem_post(&board_queue);
			i++;
		}
		sem_wait(&load);
	
		//Run around the track
		printf("Car%d-> departing\n", id);
		usleep(50000 + rand() % 500000);
		printf("Car%d-> arriving\n", id);
		
		//Unload the car
		i = 0;
		while(i < MAX_PASSENGERS) {
			sem_post(&unboard_queue);
			i++;
		}
		sem_wait(&unload);
		printf("Car%d-> unloaded\n", id);
		
		served += MAX_PASSENGERS;
	}
	
	printf("Car%d-> exit\n", id);
	pthread_exit(0);
}

//Passenger thread
void* passenger_sim(void* arg)
{
	int id = *(int*)arg;
	
	sem_wait(&board_queue);

	pthread_mutex_lock(&mx_board);
	boarders++;
	if (boarders == MAX_PASSENGERS) {
		printf("Pas%d-> last one on\n", id);
		sem_post(&load);
		boarders = 0;
	} else {
		printf("Pas%d-> getting on...\n", id);
	}
	pthread_mutex_unlock(&mx_board);
		
	sem_wait(&unboard_queue);
	
	pthread_mutex_lock(&mx_unboard);
	unboarders++;
	if (unboarders == MAX_PASSENGERS) {
		printf("Pas%d-> last one off\n", id);
		sem_post(&unload);
		unboarders = 0;
	} else {
		printf("Pas%d-> getting off...\n", id);
	}
	pthread_mutex_unlock(&mx_unboard);

	pthread_exit(0);
}


int main(int argc, char* argv[])
{
	//start timer
	struct timeval stop, start;
	gettimeofday(&start, NULL);
	
	//initializations
	boarders = 0;
	unboarders = 0;
	pthread_mutex_init(&mx_board, NULL);
	pthread_mutex_init(&mx_unboard, NULL);
	sem_init(&board_queue, 0, 0);
	sem_init(&unboard_queue, 0, 0);
	sem_init(&load, 0, 0);
	sem_init(&unload, 0, 0);
	
	//initialize the car and passengers threads
	pthread_t car;
	pthread_t passengers[NUM_PASSENGERS];
	int cid = 1;
	int pids[NUM_PASSENGERS];
	
	//start the car
	if (pthread_create(&car, NULL, car_sim, &cid) != 0)
	{
		printf("Failed to create car thread!\n");
		exit(EXIT_FAILURE);
	}
	
	int i = 0;
	//start the passengers
	while (i < NUM_PASSENGERS)
	{
		pids[i] = i+1;
		if (pthread_create(&passengers[i], NULL, passenger_sim, &pids[i]) != 0)
		{
			printf("Failed to create passenger thread!\n");
			exit(EXIT_FAILURE);
		}
		i++;
	}
	
	//wait for car to finish
	pthread_join(car, NULL);
	
	//wait for passengers to finish
	i = 0;
	while (i < NUM_PASSENGERS) {
		pthread_join(passengers[i++], NULL);
	}
	
	//destroy the locks and semaphores
	pthread_mutex_destroy(&mx_board);
	pthread_mutex_destroy(&mx_unboard);
	sem_destroy(&board_queue);
	sem_destroy(&unboard_queue);
	sem_destroy(&load);
	sem_destroy(&unload);
	
	//compute runtime
	gettimeofday(&stop, NULL);
	int elapsed = ((int)(stop.tv_sec - start.tv_sec) * 1000) + ((int)(stop.tv_usec - start.tv_usec) / 1000);
	printf("Simulation finished. Runtime: %dms\n", elapsed);
	return 0;
} 

File: roller-coaster/RollerCoaster.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;

class RollerCoaster
{
	static readonly int MAX_PASSENGERS = 4;
	static readonly int NUM_PASSENGERS = 24;

	//counters for boarding/unboarding
	static int boarders;
	static int unboarders;

	static readonly object boardLock = new object();
	static readonly object unboardLock = new object();

	static Semaphore boardQueue = new Semaphore(0, MAX_PASSENGERS);
	static Semaphore unboardQueue = new Semaphore(0, MAX_PASSENGERS);
	static Semaphore load = new Semaphore(0, 1);
	static Semaphore unload = new Semaphore(0, 1);
	
	//Car thread
	static void Car()
	{
		Random rand = new Random();
		int served = 0;
		Console.WriteLine(Thread.CurrentThread.Name + "-> start");

		while (served < NUM_PASSENGERS)
		{
			//Load up
			Console.WriteLine(Thread.CurrentThread.Name + "-> load");
			boardQueue.Release(MAX_PASSENGERS);
			load.WaitOne();

			//Run around the track
			Console.WriteLine(Thread.CurrentThread.Name + "-> departing");
			Thread.Sleep(50 + rand.Next(500));
			Console.WriteLine(Thread.CurrentThread.Name + "-> arriving");

			//Unload
			unboardQueue.Release(MAX_PASSENGERS);
			unload.WaitOne();
			Console.WriteLine(Thread.CurrentThread.Name + "-> unloaded");

			served += MAX_PASSENGERS;
		}

		Console.WriteLine(Thread.CurrentThread.Name + "-> exit");
	}

	//Passenger thread
	static void Passenger()
	{
		//wait for boarding
		boardQueue.WaitOne();

		//board
		lock(boardLock)
		{
			boarders++;
			if (boarders == MAX_PASSENGERS)
			{
				Console.WriteLine(Thread.CurrentThread.Name + "-> last one on");
				load.Release();
				boarders = 0;
			}
			else
			{
				Console.WriteLine(Thread.CurrentThread.Name + "-> getting on...");
			}
		}

		//ride running
		unboardQueue.WaitOne();

		//unboard
		lock(unboardLock)
		{
			unboarders++;
			if (unboarders == MAX_PASSENGERS)
			{
				Console.WriteLine(Thread.CurrentThread.Name + "-> last one off");
				unload.Release();
				unboarders = 0;
			}
			else
			{
				Console.WriteLine(Thread.CurrentThread.Name + "-> getting off...");
			}
		}
	}

	static void Main(string[] args)
	{
		Stopwatch stopwatch = new Stopwatch();
		stopwatch.Start();

		//start car thread
		Thread car = new Thread(Car) { Name = "Car1" };
		car.Start();

		//start passenger threads
		Thread[] passengers = new Thread[NUM_PASSENGERS];
		for(int i = 0; i < NUM_PASSENGERS; i++)
		{
			passengers[i] = new Thread(Passenger) { Name = "Pas" + (i + 1) };
			passengers[i].Start();
		}

		//wait for car to finish
		car.Join();

		//wait for passengers to finish
		for (int i = 0; i < NUM_PASSENGERS; i++)
		{
			passengers[i].Join();
		}

		Console.WriteLine("Simulation finished. Runtime: " + stopwatch.ElapsedMilliseconds + "ms");
		Console.ReadLine();
	}
}